indra/newview/llmaterialeditor.cpp

```diff
@@ -1816,58 +1821,95 @@ void LLMaterialEditor::loadLive()
+
+        if (local_mat)
+        {
+            local_material = local_mat;
+        }
+        render_material = tep->getGLTFRenderMaterial();
+    }
+    else
+    {
+        // Find an applicable material.
+        // Do this before showing message, because
+        // message is going to drop selection.
+        LLSelectedTEGetMatData func(false);
+        LLSelectMgr::getInstance()->getSelection()->applyToTEs(&func, true /*first applicable*/);
+        local_material = func.mLocalMaterial;
+        render_material = func.mMaterial;
+    }
+
+    saveMaterialAs(render_material, local_material);
+}
+
+void LLMaterialEditor::saveMaterialAs(const LLGLTFMaterial* render_material, const LLLocalGLTFMaterial *local_material)
+{
+    if (local_material)
     {
         // This is a local material, reload it from file
         // so that user won't end up with grey textures
         // on next login.
-        LLMaterialEditor::loadMaterialFromFile(func.mLocalMaterial->getFilename(), func.mLocalMaterial->getIndexInFile());
+        LLMaterialEditor::loadMaterialFromFile(local_material->getFilename(), local_material->getIndexInFile());
 
         LLMaterialEditor* me = (LLMaterialEditor*)LLFloaterReg::getInstance("material_editor");
         if (me)
         {
-            // apply differences on top
-            LLGLTFMaterial* local_mat = func.mLocalMaterial.get();
-            // don't use override mat here, it has 'hacked ids'
-            // and values, use end result.
-            LLGLTFMaterial* cmp_mat = func.mMaterial.get();
-
-            me->setBaseColor(cmp_mat->mBaseColor);
-            me->setMetalnessFactor(cmp_mat->mMetallicFactor);
-            me->setRoughnessFactor(cmp_mat->mRoughnessFactor);
-            me->setEmissiveColor(cmp_mat->mEmissiveColor);
-            me->setDoubleSided(cmp_mat->mDoubleSided);
-            me->setAlphaMode(cmp_mat->getAlphaMode());
-            me->setAlphaCutoff(cmp_mat->mAlphaCutoff);
+            // don't use override material here, it has 'hacked ids'
+            // and values, use end result, apply it on top of local.
+            me->setBaseColor(render_material->mBaseColor);
+            me->setMetalnessFactor(render_material->mMetallicFactor);
+            me->setRoughnessFactor(render_material->mRoughnessFactor);
+            me->setEmissiveColor(render_material->mEmissiveColor);
+            me->setDoubleSided(render_material->mDoubleSided);
+            me->setAlphaMode(render_material->getAlphaMode());
+            me->setAlphaCutoff(render_material->mAlphaCutoff);
 
             // most things like colors we can apply without verifying
             // but texture ids are going to be different from both, base and override
             // so only apply override id if there is actually a difference
-            if (local_mat->mBaseColorId != cmp_mat->mBaseColorId)
+            if (local_material->mBaseColorId != render_material->mBaseColorId)
```

Comment:
I think that the render material should be the correct combination of all material properties, so ideally these checks should not be needed.

Reply:
Well, actually, there is one exception: The render material has transforms, but the saved material should not.

Reply:
In this case, it's a 'file' texture, if local material's texture and render texture match, viewer needs to upload said texture, if they do not much, no upload is needed because user replaced 'file' texture with something else. Admitedly this can be handled better and I probably should do some kind of is_texture_local(render_material->mBaseColorId) even if texture doesn't match one supplied by a gltf file, but that's a low priority.